### PR TITLE
feat(windows): add state machine to background updates 💽

### PR DESCRIPTION
**Pull request by rc-swag**
Fixes: #10210

Still outstanding and noted in issue 10210

- [ ] Add error logging. 
- [ ] Implement Retry
- [ ] If called in WaitingRestart to many times abort and go back to idle
- [ ] Error log for Unable to set state for Value
- [ ] workout when we need to refresh kmcom keyboards

### CODE REVIEW

File: common/windows/delphi/general/Keyman.System.ExecuteHistory.pas [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    atom := GlobalFindAtom(AtomName);
+    if atom <> 0 then
+    begin
+      if GetLastError <> ERROR_SUCCESS then
+        RaiseLastOSError;
+
+       KL.Log('HasKeymanRun: Keyman Has Run');
+      Result := True;
+    end
+    else
```
Comment (ermshiperete) [inline common/windows/delphi/general/Keyman.System.ExecuteHistory.pas:54]: Do you want to log that Keyman hasn't run yet?

File: windows/src/desktop/kmshell/main/BackgroundUpdate.pas [hunk 2]
```diff
@@ -0,0 +1,1192 @@
...
+  SavedPath : String;
+  Filenames : TStringDynArray;
+begin
+  KL.Log('WaitingRestartState.HandleKmShell Enter');
+  // Still can't go if keyman has run
+  if HasKeymanRun then
+  begin
+    KL.Log('WaitingRestartState.HandleKmShell Keyman Has Run');
+    Result := kmShellExit;
+    // Exit; // Exit is not wokring for some reason.
```
Comment (ermshiperete) [inline windows/src/desktop/kmshell/main/BackgroundUpdate.pas:893]:
```suggestion
    // Exit; // Exit is not working for some reason.
```

File: windows/src/desktop/kmshell/main/BackgroundUpdate.pas [hunk 3]
```diff
@@ -0,0 +1,1192 @@
...
+  kmcom.Refresh;
+  kmcom.Apply;
+  System.SysUtils.DeleteFile(Package.SavePath);
+end;
+
+procedure InstallingState.DoInstallKeyman;
+var
+  s: string;
+  FResult: Boolean;
+begin
```
Comment (ermshiperete) [inline windows/src/desktop/kmshell/main/BackgroundUpdate.pas:961]: Calling `DoInstallKeyman(bucStateContext.FParams.Keyman.SavePath)` would be less WET.

File: windows/src/desktop/kmshell/main/BackgroundUpdate.pas [hunk 4]
```diff
@@ -0,0 +1,1192 @@
...
+    GetFileNamesInDirectory(SavePath, fileNames);
+    // for now we only want the exe although excute install can
+    // handle msi
+    for fileName in fileNames do
+    begin
+      fileExt := LowerCase(ExtractFileExt(fileName));
+      if fileExt = '.exe' then
+        break;
+    end;
+
```
Comment (ermshiperete) [inline windows/src/desktop/kmshell/main/BackgroundUpdate.pas:1018]: Do we have to check that we found a file with .exe extension?

File: windows/src/desktop/kmshell/main/Keyman.System.RemoteUpdateCheck.pas [hunk 5]
```diff
@@ -375,4 +218,44 @@ function TRemoteUpdateCheck.DoRun: TRemoteUpdateCheckResult;
    KL.Log(LogMessage);
  end;
 
+function ConfigCheckContinue: Boolean;
```
Comment (ermshiperete) [inline windows/src/desktop/kmshell/main/Keyman.System.RemoteUpdateCheck.pas:221]: The function name doesn't tell me much. Would `CheckForUpdatesIsDue` or `ShouldCheckForUpdates` be a better name?

File: windows/src/desktop/kmshell/main/initprog.pas [hunk 6]
```diff
@@ -384,7 +385,7 @@ procedure RunKMCOM(FMode: TKMShellMode; KeyboardFileNames: TStrings; FSilent, FF
   kdl: IKeymanDefaultLanguage;
   FIcon: string;
   FMutex: TKeymanMutex;  // I2720
-  RemoteUpdateCheck: TRemoteUpdateCheck;
+  BUpdateSM : TBackgroundUpdate;
```
Comment (ermshiperete) [inline windows/src/desktop/kmshell/main/initprog.pas:388]:
```suggestion
  BUpdateStateMachine : TBackgroundUpdate;
```
?

File: windows/src/engine/keyman/main.pas [hunk 7]
```diff
@@ -76,9 +78,13 @@ procedure RunProgram;
   hMutex: Cardinal;
 begin
 
-
+      KL.Log('Keyman RunProgram');
```
Comment (ermshiperete) [inline windows/src/engine/keyman/main.pas:81]:
```suggestion
  KL.Log('Keyman RunProgram');
```

**Comment by rc-swag** [comment]
Closing this PR as this work is being parked for at least one release.

**Pull request closed without merging**